File: backend/src/api/scenario.ts
import { Hono } from 'hono';
import { pin, diff } from '../skills/scenario/index.js';
import { getPlan, savePlan } from '../db/client.js';

export const scenarioRoute = new Hono();

scenarioRoute.post('/:id/pin', async (c) => {
  const id = c.req.param('id');
  const body = await c.req.json<{ label: string; mutation?: { ops: { path: string; op: 'set' | 'add' | 'remove'; value?: unknown; row?: unknown; id?: string }[] } }>();
  const result = await pin({ household_id: id, label: body.label, mutation: body.mutation });
  return c.json(result);
});

scenarioRoute.post('/:id/diff', async (c) => {
  const id = c.req.param('id');
  const body = await c.req.json<{ a: string; b: string }>();
  const result = await diff({ household_id: id, a: body.a, b: body.b });
  return c.json(result);
});

scenarioRoute.post('/:id/toggle', async (c) => {
  const id = c.req.param('id');
  const body = await c.req.json<{ id: string; active: boolean }>();
  const plan = await getPlan(id);
  if (!plan) return c.json({ ok: false }, 404);
  const set = new Set(plan.active_scenario_ids);
  if (body.active) set.add(body.id);
  else set.delete(body.id);
  plan.active_scenario_ids = [...set].slice(-3);
  await savePlan(plan);
  return c.json({ ok: true });
});

scenarioRoute.post('/:id/clear', async (c) => {
  const id = c.req.param('id');
  const plan = await getPlan(id);
  if (!plan) return c.json({ ok: false }, 404);
  plan.scenarios = [];
  plan.active_scenario_ids = [];
  await savePlan(plan);
  return c.json({ ok: true });
});

File: backend_py/stackwealth/api/scenario.py
"""/api/scenario — Plan A/B compare endpoints."""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..db import get_plan, save_plan
from ..skills.scenario import diff, pin

router = APIRouter()


def _json(data: Any) -> JSONResponse:
    if hasattr(data, "model_dump"):
        return JSONResponse(content=data.model_dump(mode="json"))
    return JSONResponse(content=data)


@router.post("/{id}/pin")
async def pin_scenario(id: str, request: Request) -> JSONResponse:
    body = await request.json()
    return _json(
        await pin(
            {
                "household_id": id,
                "label": body["label"],
                "mutation": body.get("mutation"),
            }
        )
    )


@router.post("/{id}/diff")
async def diff_scenarios(id: str, request: Request) -> JSONResponse:
    body = await request.json()
    return _json(await diff({"household_id": id, "a": body["a"], "b": body["b"]}))


@router.post("/{id}/toggle")
async def toggle(id: str, request: Request) -> JSONResponse:
    body = await request.json()
    plan = await get_plan(id)
    if not plan:
        return JSONResponse(content={"ok": False}, status_code=404)
    s = set(plan.active_scenario_ids)
    if body.get("active"):
        s.add(body["id"])
    else:
        s.discard(body["id"])
    plan.active_scenario_ids = list(s)[-3:]
    await save_plan(plan)
    return _json({"ok": True})


@router.post("/{id}/clear")
async def clear(id: str) -> JSONResponse:
    plan = await get_plan(id)
    if not plan:
        return JSONResponse(content={"ok": False}, status_code=404)
    plan.scenarios = []
    plan.active_scenario_ids = []
    await save_plan(plan)
    return _json({"ok": True})
